### PR TITLE
fix: LEAP-762: Fix offsets in domManager for emojis

### PR DESCRIPTION
**Pull request by Gondragos**
There were some problems with calculating offsets and applying text regions to the text/html containing emojis after domManager was introduced.
- here it might be fixed
- there are new tests for that
- RichText helper for testing is introduces

### PR fulfills these requirements
- [x] Commit message(s) and PR title follows the format `[fix|feat|ci|chore|doc]: TICKET-ID: Short description of change made` ex. `fix: DEV-XXXX: Removed inconsistent code usage causing intermittent errors`
- [x] Tests for the changes have been added/updated (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [x] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [x] Self-reviewed and ran all changes on a local instance (for bug fixes/features)

#### Change has impacts in these area(s)
- [ ] Product design
- [ ] Backend (Database)
- [ ] Backend (API)
- [x] Frontend

#### What feature flags were used to cover this change?
N/A

### Does this PR introduce a breaking change?
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [ ] No
- [x] Not sure (briefly explain the situation below)

I hope nobody would like to keep that bug.

### What level of testing was included in the change?
- [ ] e2e
- [x] integration
- [ ] unit

### Which logical domain(s) does this change affect?
`Text`, `RichText`, `HyperText`, `domManager`

### CODE REVIEW

File: web/libs/editor/src/tags/object/RichText/domManager.ts
Comment (hlomzik) [inline web/libs/editor/src/tags/object/RichText/domManager.ts:1]: Overall suggestion is to add documentation for the whole file — what are the main ideas, what is the structure and main functions here. Like "domManager keeps the internal structure of all elements inside it with offset mapping to find element by offset in O(1) and get offset of the element in zero time. Provides mappings via methods: offsetToRange, <...>, etc".

Comment (hlomzik) [reply web/libs/editor/src/tags/object/RichText/domManager.ts:1]: Also I suggest to separate string offsets and code points to understand the difference and to draw a line between methods operating one or another and to highlight the conversion points. This can be done via virtual alias types for example: `type CodePoint = number; type StringOffset = number` at least for readability, but also maybe it can really be used to typesafe the code.

**Review by hlomzik — APPROVED** [review]

File: web/libs/editor/src/tags/object/RichText/domManager.ts [hunk 1]
```diff
@@ -720,14 +788,21 @@ export default class DomManager {
...
+    let endEl = this.domData.findElementByPath(end);
 
     if (!startEl || !endEl) {
       return undefined;
     }
+    if (!(startEl instanceof DDDynamicBlock)) {
+      startEl = this.domData.findTextBlock(startEl.start, "end") as DDDynamicBlock;
+    }
+    if (!(endEl instanceof DDDynamicBlock)) {
+      // It really should be "end" and not "start" as we are looking for the exact container by the start position
```
Comment (hlomzik) [inline web/libs/editor/src/tags/object/RichText/domManager.ts:802]: this and next comments makes it even more confusing. what `startEl` and `endEl` are and why do we use `"end"` for start? `"end"` for `endEl` seems more obvious. Also why do you convert text block to dynamic block? Is it possible to define interfaces and use mostly them? like if any element in `DomData.elements` could have the same interface that would be awesome.

**Review by bmartel — APPROVED** [review]

**Review by MihajloHoma — APPROVED** [review]

File: web/libs/editor/tests/integration/e2e/ner/emoji.cy.ts [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+        expect(resultValue.globalOffsets.start).to.eq(refHyperTextMultilineResultValue.globalOffsets.start);
+        expect(resultValue.globalOffsets.end).to.eq(refHyperTextMultilineResultValue.globalOffsets.end);
+        expect(resultValue.startOffset).to.eq(refHyperTextMultilineResultValue.startOffset);
+        expect(resultValue.endOffset).to.eq(refHyperTextMultilineResultValue.endOffset);
+        expect(resultValue.text).to.eq(refHyperTextMultilineResultValue.text);
+      });
+    });
+  });
+
+  it.only('Heuristic edge case', () => {
```
Comment (Gondragos) [inline web/libs/editor/tests/integration/e2e/ner/emoji.cy.ts]:
```suggestion
  it('Heuristic edge case', () => {
```
